File: chillnote/Utils/ChecklistMarkdown.swift
import Foundation

struct ChecklistMarkdown {
    struct ParsedChecklist {
        let notes: String
        let items: [(isDone: Bool, text: String)]
    }

    private static let checkboxRegex = try? NSRegularExpression(
        pattern: #"^\s*[-*]\s*\[( |x|X)\]\s*(.*?)\s*$"#,
        options: []
    )

    static func parse(_ content: String) -> ParsedChecklist? {
        guard let checkboxRegex else { return nil }
        let lines = content.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)

        var items: [(Bool, String)] = []
        var notesLines: [String] = []
        var currentItemIndex: Int?

        for line in lines {
            let nsLine = line as NSString
            let range = NSRange(location: 0, length: nsLine.length)
            if let match = checkboxRegex.firstMatch(in: line, options: [], range: range),
               match.numberOfRanges >= 3
            {
                let doneMark = nsLine.substring(with: match.range(at: 1))
                let text = nsLine.substring(with: match.range(at: 2)).trimmingCharacters(in: .whitespacesAndNewlines)
                let isDone = doneMark.lowercased() == "x"
                items.append((isDone, text))
                currentItemIndex = items.count - 1
            } else {
                let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                if trimmed.isEmpty {
                    continue
                }

                if let currentItemIndex {
                    let existingText = items[currentItemIndex].1
                    items[currentItemIndex].1 = existingText.isEmpty
                        ? trimmed
                        : "\(existingText)\n\(trimmed)"
                } else {
                    notesLines.append(line)
                }
            }
        }

        guard !items.isEmpty else { return nil }
        let notes = notesLines.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        return ParsedChecklist(notes: notes, items: items)
    }

    static func serialize(notes: String, items: [ChecklistItem]) -> String {
        var parts: [String] = []
        let trimmedNotes = notes.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedNotes.isEmpty {
            parts.append(trimmedNotes)
            parts.append("")
        }

        for item in items.sorted(by: { $0.sortOrder < $1.sortOrder }) {
            let mark = item.isDone ? "x" : " "
            let lines = item.text
                .split(separator: "\n", omittingEmptySubsequences: false)
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }

            guard let firstLine = lines.first else {
                parts.append("- [\(mark)]")
                continue
            }

            parts.append("- [\(mark)] \(firstLine)")
            for continuation in lines.dropFirst() {
                parts.append("    \(continuation)")
            }
        }

        return parts.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func serializePlainText(notes: String, items: [ChecklistItem]) -> String {
        var parts: [String] = []
        let trimmedNotes = notes.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedNotes.isEmpty {
            parts.append(trimmedNotes)
            parts.append("")
        }

        for item in items.sorted(by: { $0.sortOrder < $1.sortOrder }) {
            // User requested no prefixes when converting back to text
            let text = item.text.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !text.isEmpty else { continue }
            parts.append(text)
        }

        return parts.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: android/app/src/main/java/com/sponteoai/chillscript/domain/ChecklistMarkdown.kt
package com.sponteoai.chillscript.domain

data class ChecklistDraftItem(val text: String, val isDone: Boolean = false)

data class ChecklistDraft(val notes: String, val items: List<ChecklistDraftItem>)

object ChecklistMarkdown {
    private val checkboxRegex = Regex("^\\s*[-*]\\s*\\[( |x|X)]\\s*(.*?)\\s*$")

    fun parse(content: String): ChecklistDraft? {
        val items = mutableListOf<ChecklistDraftItem>()
        val notesLines = mutableListOf<String>()
        var currentItemIndex: Int? = null

        content.split('\n').forEach { line ->
            val match = checkboxRegex.matchEntire(line)
            if (match != null) {
                items += ChecklistDraftItem(
                    text = match.groupValues[2].trim(),
                    isDone = match.groupValues[1].equals("x", ignoreCase = true),
                )
                currentItemIndex = items.lastIndex
            } else {
                val trimmed = line.trim()
                if (trimmed.isNotEmpty()) {
                    val itemIndex = currentItemIndex
                    if (itemIndex == null) {
                        notesLines += line
                    } else {
                        val existing = items[itemIndex]
                        items[itemIndex] = existing.copy(
                            text = if (existing.text.isEmpty()) trimmed else "${existing.text}\n$trimmed",
                        )
                    }
                }
            }
        }

        if (items.isEmpty()) return null
        return ChecklistDraft(notesLines.joinToString("\n").trim(), items)
    }

    fun serialize(notes: String, items: List<ChecklistDraftItem>): String {
        val parts = mutableListOf<String>()
        notes.trim().takeIf { it.isNotEmpty() }?.let {
            parts += it
            parts += ""
        }
        items.forEach { item ->
            val mark = if (item.isDone) "x" else " "
            val lines = item.text.split('\n').map(String::trim).filter(String::isNotEmpty)
            if (lines.isEmpty()) {
                parts += "- [$mark]"
            } else {
                parts += "- [$mark] ${lines.first()}"
                lines.drop(1).forEach { parts += "    $it" }
            }
        }
        return parts.joinToString("\n").trim()
    }

    fun serializePlainText(notes: String, items: List<ChecklistDraftItem>): String = buildList {
        notes.trim().takeIf(String::isNotEmpty)?.let {
            add(it)
            add("")
        }
        items.map { it.text.trim() }.filter(String::isNotEmpty).forEach(::add)
    }.joinToString("\n").trim()
}
